lib.types.unique: Check inner type deeply

roberth opened the pull request:
## Description of changes

Provide a safe alternative to `uniq` because of
- #284381 

This doesn't change uniq. Why not?

- In NixOS it seems that uniq is mostly used with simple types that are fully checked by t.check.

- It exists for much longer and is used more widely.

- I believe we should deprecate it, because unique was already better.

- `unique` can be a proving ground, and `uniq` can be changed, deprecated or whatnot later.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
lib.types.unique: Check inner type deeply

This doesn't change uniq. Why not?

- In NixOS it seems that uniq is only used with
  simple types that are fully checked by t.check.

- It exists for much longer and is used more widely.

- I believe we should deprecate it, because unique was
  already better.

- unique can be a proving ground.

roberth committed:
doc/option-types: Definitions are not declared

infinisil commented on lib/types.nix:
```diff
@@ -629,7 +629,7 @@ rec {
       name = "unique";
       inherit (type) description descriptionClass check;
-      merge = mergeUniqueOption { inherit message; };
+      merge = mergeUniqueOption { inherit message; inherit (type) merge; };
```
This feels like a small layer violation to me, `lib/options.nix` should only _provide` merge functions, not make use of them.

Instead we could do this:
```suggestion
      merge =
        if length defs == 1 then
          # Makes sure that nested values are also type-checked
          type.merge loc defs
        else
          throw "The option `${showOption loc}' is defined multiple times while it's expected to be unique.\n${message}\nDefinition values:${showDefs defs}\n${prioritySuggestion}";
```

roberth replied:
I don't think there was any real layering to begin with. Options _use_ types, but types _use_ `options.nix` to implement merging.

Also I find it surprising that you object to the construction of merge functions from other merge functions.
Constructing \<x> from other \<x> is exactly what makes languages and libraries scale.

infinisil replied:
Fair, just felt a bit like that, but I don't really mind that. Though I do like the simplicity of the suggestion still :)

infinisil commented on nixos/doc/manual/development/option-types.section.md:
```diff
@@ -326,7 +326,14 @@ Composed types are types that take a type as parameter. `listOf
+    ::: {.warning}
+    Deep type checks will not be performed. If you need `t` to check attribute
+    values or list items, use `types.unique` instead.
+    A future version of the module system may change `types.uniq` to match
+    `types.unique`'s behavior; to also check those lazy substructures.
```
Not a fan of intentionally different behavior for two functions that originally had the same behavior. I don't think there's a good reason these should be distinguished like that.

I'd be in favor of making both `types.uniq` and `types.unique` behave in the stricter way. Really it not being strict before should be considered a bug.

roberth replied:
> making both `types.uniq` and `types.unique` behave in the stricter way. Really it not being strict before should be considered a bug.

Done and agreed.

roberth committed:
lib.types.uniq: Check inner type

We now reuse the `unique` type, which implements this.
Keeping the duplication around would be bad at this point.

roberth committed:
lib.option.mergeUniqueOption: Simplify and add warning about merge function

The previous code was optimized for the old uniq behavior, which did not
call merge. That's changed, so the legacy path is not a hot path anymore,
and is not worth any tech debt.

infinisil commented:
I don't think this should be backported since it's potentially backwards incompatible

roberth commented:
I agree. The initial version - when I added the label - was far less breaking.
This should not be backported, even if it could be considered a bugfix.
